### PR TITLE
Initial work for Dropwizard integration.

### PR DESCRIPTION
**Pull request by sduskis**
- Adds a wrapper for Timer and Counter
- Adds RPC and Operation level timers, and retry counters.

This should start towards the goals of #505.

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/BigtableClientMetrics.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  private static MetricRegistry CLIENT_STATS;
+
+  private static final Counter NULL_COUNTER = new Counter() {
+    @Override
+    public void inc() {
+    }
+  };
+
+  private static final Timer NULL_TIMER = new Timer() {
+
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/BigtableClientMetrics.java]: Kill the newline

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/BigtableClientMetrics.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+          Slf4jReporter.forRegistry(CLIENT_STATS)
+              .outputTo(logger)
+              .convertRatesTo(TimeUnit.SECONDS)
+              .convertDurationsTo(TimeUnit.MILLISECONDS)
+              .build();
+      reporter.start(1, TimeUnit.MINUTES);
+    }
+  }
+
+  public static void enable() {
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/BigtableClientMetrics.java]: Might as well synchronize, unless this is being called somewhere safe that I missed.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/Counter.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.metrics;
+
+/**
+ * Wrapper around a Dropwizard Metrics {@link com.codahale.metrics.Counter}.
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/Counter.java]: No need to mention the Dropwizard implementation in the interface comments

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/BigtableClientMetrics.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    Logger logger = LoggerFactory.getLogger(BigtableClientMetrics.class);
+    if (logger.isTraceEnabled()) {
+      CLIENT_STATS = new MetricRegistry();
+      final Slf4jReporter reporter =
+          Slf4jReporter.forRegistry(CLIENT_STATS)
+              .outputTo(logger)
+              .convertRatesTo(TimeUnit.SECONDS)
+              .convertDurationsTo(TimeUnit.MILLISECONDS)
+              .build();
+      reporter.start(1, TimeUnit.MINUTES);
```
Comment (garye) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/metrics/BigtableClientMetrics.java]: Maybe make this configurable?

**Comment by sduskis** [comment]
I updated the comments for Timer, Counter and BigtableClientMetrics.

In terms of configurability, I was thinking of programmatic configuration.  The Metrics is a singleton thing, but most of the configuration we have is Connection level.   Users can customize reporters programmatically or get the out of the box simple implementation.  Do you have a better alternative?

**Comment by garye** [comment]
I think programmatic configurability makes sense. It seemed like this default reporter could be potentially be more or less verbose than someone wanted and I wanted to make sure that anyone who wanted tracing didn't also have to stick with the once-per-minute log dump.

**Comment by garye** [comment]
LGTM - I think I missed that just enabling that certain logger for tracing did all of this. That seems ok for now.

Perhaps could also add some pre-canned reporters that are accessible via static methods on BigtableClientMetrics as a middle ground between unconfigurable tracing and doing whatever you want with the metric API. Like newLoggingReporter(TimeUnit reportingInterval), newConsoleReporter(TimeUnit reportingInterval), etc. 

Something to think about for next revision, though.

**Comment by sduskis** [comment]
Thanks for the review.  I like those suggestions.  Lets discuss the options in more detail.

**Pull request merged**